### ISSUE TITLE
1.0: libwacom is not LTO ready

### ISSUE THREAD

**Issue by kloczek** [linuxwacom/libwacom#152]
When compiled with LTO optimisation linkiong of the test suite fails:

<pre>[tkloczko@barrel test]$ make
/bin/sh ../libtool  --tag=CC   --mode=link gcc  -O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fstack-protector-strong -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -flto -Os -no-install -Wl,-z,relro -Wl,--as-needed  -Wl,-z,now -specs=/usr/lib/rpm/redhat/redhat-hardened-ld -flto -fuse-linker-plugin -o test-deprecated test-deprecated.o ../libwacom/libwacom.la  -ldl 
libtool: link: gcc -O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fstack-protector-strong -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -flto -Os -Wl,-z -Wl,relro -Wl,--as-needed -Wl,-z -Wl,now -specs=/usr/lib/rpm/redhat/redhat-hardened-ld -flto -fuse-linker-plugin -o test-deprecated test-deprecated.o  ../libwacom/.libs/libwacom.so -lgudev-1.0 -lgobject-2.0 -lglib-2.0 -ldl -Wl,-rpath -Wl,/home/tkloczko/rpmbuild/BUILD/libwacom-0.99.901/libwacom/.libs
/usr/bin/ld: /tmp/test-deprecated.thrYTc.ltrans0.ltrans.o: in function `main&apos;:
/home/tkloczko/rpmbuild/BUILD/libwacom-0.99.901/test/test-deprecated.c:77: undefined reference to `libwacom_match_destroy@LIBWACOM_0.33&apos;
/usr/bin/ld: /home/tkloczko/rpmbuild/BUILD/libwacom-0.99.901/test/test-deprecated.c:78: undefined reference to `libwacom_match_new@LIBWACOM_0.33&apos;
/usr/bin/ld: /home/tkloczko/rpmbuild/BUILD/libwacom-0.99.901/test/test-deprecated.c:79: undefined reference to `libwacom_error_set@LIBWACOM_0.33&apos;
/usr/bin/ld: /home/tkloczko/rpmbuild/BUILD/libwacom-0.99.901/test/test-deprecated.c:80: undefined reference to `libwacom_stylus_destroy@LIBWACOM_0.33&apos;
/usr/bin/ld: /home/tkloczko/rpmbuild/BUILD/libwacom-0.99.901/test/test-deprecated.c:81: undefined reference to `libwacom_update_match@LIBWACOM_0.33&apos;
collect2: error: ld returned 1 exit status
make: *** [Makefile:664: test-deprecated] Error 1
make: Target &apos;all&apos; not remade because of errors.
</pre>

**Comment by kloczek** [comment]
The same is when libwacom is build using meson.

**Comment by whot** [comment]
Any suggestions on how to fix this? this test is largely to verify that the symbols are still in the library (through `dlsym` and `asm` mappings) so I *think* this is fine to fail under LTO and just needs to be disabled when LTO is active. Not 100% sure though where LTO kicks in here.

**Comment by kloczek** [comment]
Please have look on other simillar case
https://github.com/SELinuxProject/selinux/issues/169

**Comment by whot** [comment]
thanks, but I admit it doesn't give me enough details to continue. I managed to reproduce with `export CC="gcc -flto"` before meson, so it's confirmed at least. But looking at the `libwacom.so` now shows:
```
000000000000971e t libwacom_error_set@LIBWACOM_0.33
00000000000096fc t libwacom_match_destroy@LIBWACOM_0.33
0000000000009706 t libwacom_match_new@LIBWACOM_0.33
0000000000009775 t libwacom_stylus_destroy@LIBWACOM_0.33
000000000000977f t libwacom_update_match@LIBWACOM_0.33
```

So these functions are now local. This indicates that the LTO modifies the the `.so` file, it's not just an issue in the test. This would be an ABI break which is something I tried to avoid with the whole symbol dance here.

So - what's going on and how do I fix this?
